sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java

```diff
@@ -104,38 +115,60 @@ public void encode(Result<DestinationT> value, OutputStream outStream, Context c
     public void verifyDeterministic() {}
   }
 
-  WriteBundlesToFiles(String tempFilePrefix) {
+  WriteBundlesToFiles(String tempFilePrefix,
+                      TupleTag<KV<ShardedKey<DestinationT>, TableRow>> unwrittedRecordsTag,
+                      int maxNumWriters) {
     this.tempFilePrefix = tempFilePrefix;
+    this.unwrittedRecordsTag = unwrittedRecordsTag;
+    this.maxNumWriters = maxNumWriters;
   }
 
   @StartBundle
   public void startBundle(Context c) {
-    // This must be done each bundle, as by default the {@link DoFn} might be reused between
+    // This must be done for each bundle, as by default the {@link DoFn} might be reused between
     // bundles.
     this.writers = Maps.newHashMap();
   }
 
   @ProcessElement
   public void processElement(ProcessContext c) throws Exception {
     TableRowWriter writer = writers.get(c.element().getKey());
+    if (writer != null && writer.getByteSize() > Write.MAX_FILE_SIZE) {
+      // File is too big. Close it and open a new file.
+      TableRowWriter.Result result = writer.close();
+      c.output(new Result<>(result.resourceId.toString(), result.byteSize, c.element().getKey()));
+      writer = null;
+    }
     if (writer == null) {
-      writer = new TableRowWriter(tempFilePrefix);
-      writer.open(UUID.randomUUID().toString());
-      writers.put(c.element().getKey(), writer);
-      LOG.debug("Done opening writer {}", writer);
+      // Only create a new writer if we have fewer than maxNumWriters already in this bundle.
+      if (writers.size() <= maxNumWriters) {
+        writer = new TableRowWriter(tempFilePrefix);
+        writer.open(UUID.randomUUID().toString());
+        writers.put(c.element().getKey(), writer);
+        LOG.debug("Done opening writer {}", writer);
+      }
     }
-    try {
-      writer.write(c.element().getValue());
-    } catch (Exception e) {
-      // Discard write result and close the write.
+    if (writer != null) {
       try {
-        writer.close();
-        // The writer does not need to be reset, as this DoFn cannot be reused.
-      } catch (Exception closeException) {
-        // Do not mask the exception that caused the write to fail.
-        e.addSuppressed(closeException);
+        writer.write(c.element().getValue());
+      } catch (Exception e) {
+        // Discard write result and close the write.
+        try {
+          writer.close();
```

Comment:
Why do we need to do this here, but don't need to do it in a similar place in WriteGroupedRecordsToFiles?